File: domain/src/main/java/com/catalog/domain/category/CategoryGateway.java
package com.catalog.domain.category;

import com.catalog.domain.pagination.SearchQuery;
import com.catalog.domain.pagination.Pagination;

import java.util.Optional;

public interface CategoryGateway {
    Category create(Category aCategory);
    Optional<Category> findById(CategoryID anId);
    Category update(Category aCategory);
    void deleteById(CategoryID anId);
    Pagination<Category> findAll(SearchQuery aQuery);
}

File: domain/src/main/kotlin/category/CategoryGateway.kt
package category

import pagination.Pagination

interface CategoryGateway {
    fun create(aCategory: Category): Category

    fun deleteById(anId: CategoryID)

    fun findById(anId: CategoryID): Category?

    fun update(aCategory: Category): Category

    fun findAll(aQuery: CategorySearchQuery): Pagination<Category>
}